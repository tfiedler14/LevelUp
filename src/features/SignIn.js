import React, { useEffect, useState } from 'react';
import { connect } from 'react-redux';
import { Text, View, ImageBackground, Image } from 'react-native';
import { setErrors } from '../logic/errors/actions';
import { setLocation } from '../logic/location/actions';
import { setAuth } from '../logic/auth/actions';
import { firebaseApp } from '../../Const';
import { compose } from 'redux';
import { Field, getFormValues, initialize, reduxForm } from 'redux-form';
import { WrappedTextInput } from '../shared-components/FormField';
import { Button, Card } from 'react-native-paper';
import EStyleSheet from 'react-native-extended-stylesheet';
import { FormHeader } from '../shared-components/FormHeader';
import { AppLoading } from 'expo';

export const SignIn = ({ setLocation, handleSubmit, setAuthentication, setErrors, errors }) => {
  return (
    <> 
    <ImageBackground source={require('../../assets/images/newBackground.png')} style={{height: '100%', width: '100%'}}>
      {errors.signIn && (
        <Card style={styles.errorCard}>
          <View>
            <Text style={{color: '#FFFFFF'}}>Incorrect email and password combination</Text>
          </View>
        </Card>
      )}
      <View>
      <Image
    style={{alignSelf: 'center', resizeMode: 'contain', width: '92%'}}
    source={require('../../assets/images/newLogo.png')}
  />
      <Card style={styles.card}>
        <View>
          <FormHeader title={'Sign In'}  />
          <Field
            name="email"
            autoCapitalize='none'
            id="email"
            props={{ title: 'Email', textContentType: 'emailAddress', autoCompleteType: 'email' }}
            component={WrappedTextInput} />

              <Field
                name="password"
                id="password"
                props={{
                  title: 'Password',
                  textContentType: 'password',
                  autoCompleteType: 'password',
                  password: true
                }}
                component={WrappedTextInput}
              />
              <View style={styles.buttons}>
                <Button
                  color="#edbf18"
                  uppercase={false}
                  mode="contained"
                  onPress={handleSubmit(values => {
                    handleLogin(values, setLocation, setAuthentication, setErrors);
                  })}>
                  Sign In
                </Button>
                <Button
                  color="white"
                  uppercase={false}
                  mode="text"
                  onPress={() => setLocation('signup')}>
                  Need an account? Sign up!
                </Button>
              </View>
            </View>
          </Card>
        </View>
      </ImageBackground>
    </>
  );
};

const handleLogin = (values, setLocation, setAuth, setErrors) => {
  firebaseApp
    .auth()
    .signInWithEmailAndPassword(values.email, values.password)
    .then(response => {
      console.log(response);
      setAuth({ loggedIn: true, email: response.user.email, uid: response.user.uid });
      setLocation('calendar');
    })
    .catch(error => {
      console.log('Failed to sign in.');
      setErrors({ signIn: true });
    });
};

const styles = EStyleSheet.create({
  card: {
    padding: '1rem',
    margin: '1rem',
    backgroundColor: '#555'
    //  border: 'none',
  },
  mainView: {
    paddingTop: '4rem'
  },
  errorCard: {
    padding: '1rem',
    margin: '1rem',
    backgroundColor: '#770000'
  },

  buttons: {
    marginTop: '1rem',
    marginLeft: '2rem',
    marginRight: '2rem'
  }
});

export const mapStateToProps = state => {
  console.log(state);
  return {
    location: state.location,
    values: getFormValues('sign-in-form')(state),
    errors: state.errors,
    initialValues: {
      email: 'tf@tf.com',
      password: 'tftftf'
    }
  };
};

export const mapDispatchToProps = dispatch => {
  return {
    setAuthentication: auth => {
      dispatch(setAuth(auth));
    },
    setLocation: location => {
      dispatch(setLocation(location));
    },
    setErrors: errors => {
      dispatch(setErrors(errors));
    }
  };
};

export const validate = values => {
  const errors = {};

  if (!values.email) {
    errors.email = 'Required';
  } else if (!/^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,4}$/i.test(values.email)) {
    errors.email = 'Invalid email address';
  }

  if (!values.password) {
    errors.password = 'Required';
  } else if (values.password.length < 6 || values.password.length > 16) {
    errors.password = 'Must be 6-16 characters';
  }

  return errors;
};

export default compose(
  connect(
    mapStateToProps,
    mapDispatchToProps
  ),
  reduxForm({ form: 'sign-in-form', validate, enableReinitialize: true }),
)(SignIn);
