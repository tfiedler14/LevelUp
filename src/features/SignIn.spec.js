import React from 'react';
import { shallow } from 'enzyme';
import { mapDispatchToProps, mapStateToProps, SignIn, validate } from './SignIn';

describe('SignIn', () => {
  it('should render sign in correctly', () => {
    const component = shallow(
      <SignIn
        setLocation={jest.fn()}
        handleSubmit={jest.fn()}
        setAuthentication={jest.fn()}
        initialize={jest.fn()}
        setErrors={jest.fn()}
        errors={[]}
        values={[]}
      />
    );

    expect(component).toMatchSnapshot();
  });

  it('should render sign in correctly with errors', () => {
    const component = shallow(
      <SignIn
        setLocation={jest.fn()}
        handleSubmit={jest.fn()}
        setAuthentication={jest.fn()}
        initialize={jest.fn()}
        setErrors={jest.fn()}
        errors={{signIn: "the app is on fire"}}
        values={[]}
      />
    );

    expect(component).toMatchSnapshot();
  });
});

describe('mapStateToProps', () => {
  it('should return the correct auth, profile, and initialValues', () => {
    const state = {
      location: 'signin',
      errors: { signin: false },
      form: {
        'sign-in-form': {
          values: {
            test: 'hi'
          }
        }
      }
    };

    expect(mapStateToProps(state, { editProp: false })).toEqual({
      location: 'signin',
      errors: { signin: false },
      values: { test: 'hi' }
    });
  });
});

describe('validatorTest', () => {
  const dummyValues = {
    email: 'test@gmail.com',
    password: 'password test'
  };

  it('should validate email is required', () => {
    expect(validate({ ...dummyValues, email: null })).toEqual({
      email: 'Required'
    });
  });

  it('should validate real email', () => {
    expect(
      validate({ ...dummyValues, email: 'the quick brown fox jumps over the lazy dog' })
    ).toEqual({
      email: 'Invalid email address'
    });
  });

  it('should validate password is required', () => {
    expect(validate({ ...dummyValues, password: null })).toEqual({
      password: 'Required'
    });
  });

  it('should validate password is greater than 6 characters', () => {
    expect(
      validate({
        ...dummyValues,
        password: 'hi'
      })
    ).toEqual({
      password: 'Must be 6-16 characters'
    });
  });

  it('should validate password is less than 16 characters', () => {
    expect(
      validate({
        ...dummyValues,
        password: 'the quick brown fox jumps over the lazy dog'
      })
    ).toEqual({
      password: 'Must be 6-16 characters'
    });
  });
});

describe('mapDispatchToProps', () => {
  it('should dispatch setAuthentication when setAuthentication is called', () => {
    const dispatch = jest.fn();
    const props = mapDispatchToProps(dispatch);
    props.setAuthentication('home');

    expect(dispatch).toHaveBeenCalledWith({ auth: 'home', type: 'SET_AUTH' });
  });

  it('should dispatch setErrors when setErrors is called', () => {
    const dispatch = jest.fn();
    const props = mapDispatchToProps(dispatch);
    props.setErrors({});

    expect(dispatch).toHaveBeenCalledWith({ errors: {}, type: 'SET_ERRORS'});
  });

  it('should dispatch setLocation when setLocation is called', () => {
    const dispatch = jest.fn();
    const props = mapDispatchToProps(dispatch);
    props.setLocation('home');

    expect(dispatch).toHaveBeenCalledWith({ location: 'home', type: 'SET_LOCATION' });
  });
});
