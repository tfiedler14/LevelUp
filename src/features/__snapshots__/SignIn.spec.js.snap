// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignIn should render sign in correctly 1`] = `
<Fragment>
  <ImageBackground
    source={1}
    style={
      Object {
        "height": "100%",
        "width": "100%",
      }
    }
  >
    <View
      style={
        Object {
          "paddingTop": 150,
        }
      }
    >
      <Image
        source={1}
        style={
          Object {
            "alignSelf": "center",
            "resizeMode": "contain",
            "width": "92%",
          }
        }
      />
      <withTheme(Card)>
        <View>
          <FormHeader
            title="Sign In"
          />
          <Field
            component={[Function]}
            id="email"
            name="email"
            props={
              Object {
                "autoCompleteType": "email",
                "textContentType": "emailAddress",
                "title": "Email",
              }
            }
          />
          <Field
            component={[Function]}
            id="password"
            name="password"
            props={
              Object {
                "autoCompleteType": "password",
                "password": true,
                "textContentType": "password",
                "title": "Password",
              }
            }
          />
          <View>
            <withTheme(Button)
              color="#cda845"
              mode="contained"
              uppercase={false}
            >
              Sign In
            </withTheme(Button)>
            <withTheme(Button)
              color="white"
              mode="text"
              onPress={[Function]}
              uppercase={false}
            >
              Need an account? Sign up!
            </withTheme(Button)>
          </View>
        </View>
      </withTheme(Card)>
    </View>
  </ImageBackground>
</Fragment>
`;

exports[`SignIn should render sign in correctly with errors 1`] = `
<Fragment>
  <ImageBackground
    source={1}
    style={
      Object {
        "height": "100%",
        "width": "100%",
      }
    }
  >
    <withTheme(Card)>
      <View>
        <Text
          style={
            Object {
              "color": "#FFFFFF",
            }
          }
        >
          Incorrect email and password combination
        </Text>
      </View>
    </withTheme(Card)>
    <View
      style={
        Object {
          "paddingTop": 150,
        }
      }
    >
      <Image
        source={1}
        style={
          Object {
            "alignSelf": "center",
            "resizeMode": "contain",
            "width": "92%",
          }
        }
      />
      <withTheme(Card)>
        <View>
          <FormHeader
            title="Sign In"
          />
          <Field
            component={[Function]}
            id="email"
            name="email"
            props={
              Object {
                "autoCompleteType": "email",
                "textContentType": "emailAddress",
                "title": "Email",
              }
            }
          />
          <Field
            component={[Function]}
            id="password"
            name="password"
            props={
              Object {
                "autoCompleteType": "password",
                "password": true,
                "textContentType": "password",
                "title": "Password",
              }
            }
          />
          <View>
            <withTheme(Button)
              color="#cda845"
              mode="contained"
              uppercase={false}
            >
              Sign In
            </withTheme(Button)>
            <withTheme(Button)
              color="white"
              mode="text"
              onPress={[Function]}
              uppercase={false}
            >
              Need an account? Sign up!
            </withTheme(Button)>
          </View>
        </View>
      </withTheme(Card)>
    </View>
  </ImageBackground>
</Fragment>
`;
